File: 2015/python/13.py
# Advent of Code 2015, day 13
# (c) aichingert

import itertools

def solve(ppl,hap):
    ans = 0

    for arr in itertools.permutations(ppl):
        cur = 0
        for i in range(len(arr)-1):
            cur += hap[(arr[i],arr[i+1])] + hap[(arr[i+1],arr[i])] 
        ans = max(ans, cur + hap[(arr[-1],arr[0])] + hap[(arr[0],arr[-1])])
    return ans

ppl = set()
hap = dict()

for l in open("../input/13").read().split('\n'):
    vls = l.split(' ')
    h = int(vls[3])
    if vls[2] == 'lose':
        h *= -1
    ppl.add(vls[0])
    hap[(vls[0],vls[-1][0:len(vls[-1])-1])] = h

print("Part 1:", solve(ppl,hap))

for p in ppl:
    hap[(p, "you")] = 0
    hap[("you", p)] = 0
ppl.add("you")

print("Part 2:", solve(ppl,hap))

File: 2015/rust/13.rs
// Advent of Code 2015, day 13
// (c) aichingert

use std::collections::{HashMap, HashSet};
#[path="../../utils/rust/permutations.rs"] mod permutations;
use permutations::permutations;

fn solve(hap: &HashMap<(String, String), i32>, ppl: &HashSet<String>) -> i32 {
    let mut ans = 0;
    let mut perms: Vec<Vec<String>> = Vec::new();
    permutations(ppl.len(), &mut ppl.iter().map(|p| p.clone()).collect::<Vec<String>>(), &mut perms);

    for perm in perms {
        let mut cur = 0;

        for i in 0..perm.len()-1 {
            cur += hap[&(perm[i].clone(), perm[i+1].clone())];
            cur += hap[&(perm[i+1].clone(), perm[i].clone())];
        }
        ans = ans.max(cur + hap[&(perm[0].clone(), perm[ppl.len()-1].clone())] + hap[&(perm[ppl.len()-1].clone(), perm[0].clone())])
    }

    ans
}

fn parse() -> (HashMap<(String, String), i32>, HashSet<String>) {
    let mut happiness: HashMap<(String,String), i32> = HashMap::new();
    let mut people: HashSet<String> = HashSet::new();

    for line in std::fs::read_to_string("../input/13").unwrap().lines() {
        let vls = line.split(' ').collect::<Vec<&str>>();

        let hap = match vls[2] {
            "gain" => vls[3].parse::<i32>().unwrap(),
            _ => vls[3].parse::<i32>().unwrap()*(-1),
        };

        happiness.insert((vls[0].to_string(), vls[10][..vls[10].len()-1].to_string()), hap);
        people.insert(vls[0].to_string());
    }

    (happiness, people)
}

fn main() {
    let (mut hap, mut ppl) = parse();

    println!("Part 1: {}", solve(&hap, &ppl));
    for p in ppl.iter() {
        hap.insert((p.clone(), "me".to_string()), 0);
        hap.insert(("me".to_string(), p.clone()), 0);
    }
    ppl.insert("me".to_string());
    println!("Part 2: {}", solve(&hap, &ppl));
}
